Prove that `compile` never fails on well-typed expressions

zhengyao-lin opened the pull request:
This PR adds a theorem stating that `compile` never fails and results in a `Term` with the corresponding type when the input is a `TypedExpr`.
This was not verified before, and there was in fact an issue causing SymCC to fail on validated policies.

The exact statement is
```
theorem compile_well_typed {Γ : Environment} {εnv : SymEnv} {tx : TypedExpr} :
  εnv = SymEnv.ofEnv Γ →
  TypedExpr.WellTyped Γ tx →
  εnv.WellFormedFor tx.toExpr →
  ∃ t : Term,
    compile tx.toExpr εnv = .ok t ∧
    t.typeOf = .option (TermType.ofType tx.typeOf)
```

Ideally `εnv.WellFormedFor tx.toExpr` should also be derived from `εnv = SymEnv.ofEnv Γ` and `TypedExpr.WellTyped Γ tx`, but I'm hoping to merge this version first and get some feedback.

Thanks in advance for any suggestions! and please be critical to my Lean proof style since I'm new to Lean :)

cdisselkoen commented on cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean:
```diff
@@ -0,0 +1,2231 @@
+  case prim _ h _ _ =>
+    unfold TermType.cedarType? at *
+    split at h
+    any_goals simp at h
+    any_goals contradiction
```
here and in some other spots, I think it would be preferable to use `·` or explicit case names rather than `any_goals`, particularly if the tactic you're using only applies to one goal anyway.  If something changes in our definitions, it will be easier to figure out what went wrong if the proof contains an explicit record of which tactic was "supposed" to apply to which goal.

cdisselkoen commented on cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean:
```diff
@@ -0,0 +1,2231 @@
+
+  unfold SymEnv.WellFormedFor SymEnv.WellFormed SymRequest.WellFormed at hwf
+
+  rcases hwf with ⟨⟨⟨_, hprincipal, _, haction, _, hresource, _, hcontext⟩, _⟩, _⟩
```
here's another case where I'd use `have` instead of `rcases`

cdisselkoen reviewed (approved):
Very nice result, only minor comments

zhengyao-lin commented on cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean:
```diff
@@ -0,0 +1,2231 @@
+a particular entity type, when `SymEnv.ofEnv Γ` must also
+have tags for it
+-/
+theorem SymEnv_of_preserves_tags
+  {Γ : Environment} {ety : EntityType} {ty : CedarType}
+  (h : Γ.ets.tags? ety = some (some ty)) :
+  ∃ τags : SymTags,
+    (SymEnv.ofEnv Γ).entities.tags ety = τags ∧
+    τags.vals.outType = TermType.ofType ty
```
I think that would still hold. Although in the way it's used, we also need the fact that `Γ.ets.tags? ety = some (some ty)` implies `(SymEnv.ofEnv Γ).entities.tags ety = some (some τags)`; but I guess I can prove that in a separate theorem.

emina reviewed (changes requested):
Nice work!

I've left comments to help iterate on the proof :)

The high-level bits are:

* Use `simp only` instead of `simp` unless you're closing a goal.  This helps with proof stability.  (Lean can help you make this change; see comments.)

* Use `rcases` only for desctructuring disjunctions and `have` for everything else.

* Break up long proofs into smaller lemmas.  The rule of thumb is that individual proofs shouldn't be longer than, say, 200 lines of code or so (similar to function bodies).

* The proof has re-proved some lemmas we already have.  Let's delete those and replace with existing lemmas (see comments).

zhengyao-lin commented on cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean:
```diff
@@ -0,0 +1,2209 @@
+        apply List.sizeOf_snd_lt_sizeOf_list
+        exact hx
+
+/--
```
oh I totally missed this. It's interesting how they are exactly the same lol

zhengyao-lin commented:
@cdisselkoen @emina Thank you so much for your feedback!

I have gone through the entire proof again and addressed most of your suggestions.
In particular,
- I've replaced all non-final `simp` with `simp only`
- Refactored the proof a bit to reduce proof size; I think all theorems are less than ~150 lines now
- I tried expanding some of the `any_goals` into cases (except for some simple goals such as for `CompileWellTypedCondition.eliminate_*`, or some cases where `any_goals assumption` helps with unification of lemma application)
- All empty lines in proofs are removed
- `MapListValueRelation` and related lemmas are replaced with `Forallᵥ` or `Forall₂`
- `rcases => have`

A few general Lean questions:
- Why is `simp` ok as the final tactic? Does Lean guarantee that `simp` only gets more/stronger rewrites over updates?
- I'm a bit new to the Lean build system, how do I guarantee that `Cedar.Thm.SymCC.Compiler.WellTyped` gets verified with `lake build`? Right now it seems to skip it unless some module imports it

cdisselkoen commented on cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean:
```diff
@@ -0,0 +1,2209 @@
+    simp [hcomp_x]
+
+  -- Compilation of all fields succeeds (the exact version)
+  have hcomp_xs :
```
Nice find!

shaobo-he-aws commented on cedar-lean/Cedar/Thm/Data/List/Canonical.lean:
```diff
@@ -46,6 +46,44 @@ theorem insertCanonical_not_nil [DecidableEq β] [LT β] [DecidableLT β] (f : 
     split at h <;> try trivial
     split at h <;> trivial
 
+theorem mem_insertCanonical [LT β] [Cedar.Data.StrictLT β] [DecidableLT β]
```
Nit: rename it to `insertCanonical_mem`.

shaobo-he-aws commented on cedar-lean/Cedar/Thm/Data/List/Canonical.lean:
```diff
@@ -46,6 +46,44 @@ theorem insertCanonical_not_nil [DecidableEq β] [LT β] [DecidableLT β] (f : 
+    split; any_goals simp
+    apply Or.inr ih
+
+theorem insertCanonical_new
```
We should use a more descriptive name here.

zhengyao-lin replied:
how about `insertCanonical_new_element`?

shaobo-he-aws replied:
It's kind of tricky. I was thinking along the line of something like `insertCanonical_preverses_non_duplicate_element`. Feel free to ignore it if you feel it's too complicated.

zhengyao-lin replied:
Sure! I don't have a strong opinion on naming

cdisselkoen commented on cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean:
```diff
@@ -0,0 +1,1628 @@
+/--
+A sufficient condition for `CompileWellTyped` to hold.
+-/
+def CompileWellTypedCondition (tx : TypedExpr) (Γ : Environment) (εnv : SymEnv) : Prop :=
+  εnv = SymEnv.ofEnv Γ ∧
+  TypedExpr.WellTyped Γ tx ∧
+  εnv.WellFormedFor tx.toExpr
```
Could this be refactored to not require `εnv` as a parameter, and replace all usages of it with `SymEnv.ofEnv Γ`?  This would improve ergonomics and I think clarity?  (Open to arguments why this would not improve clarity)

zhengyao-lin replied:
I was trying this change, but I realized that a lot of the theorems have the `εnv` parameter and rephrasing them in term of `SymEnv.ofEnv Γ` might make them a bit complicated.

Also this definition is likely subject to change once we prove `WellFormedFor` from the first two conditions. So maybe we can simplify things then?

cdisselkoen replied:
Sounds good to me.
